[TN] React Context and Apollo Client (#356)
apollographql/docs: src/content/technotes/TN0034-react-context.mdx

---
title: React Context with Apollo Client
id: TN0034
tags: [client]
---

> 
> 

Reviewer: ```suggestion

```

> React Context provides a way to pass data through the component tree without having to pass props down manually at every level in your application which is also known as props-drilling.

Reviewer: This feels wordy but not really sure of a good way to reword this.

Props-drilling can cause problems when you need to rename a property or update the data type and can cause bloat within your application as each component in the tree needs to be aware of properties they may not be using. Instead our data will be accessible through a React Hook, which can be called anywhere in the nested component tree.

> The following code defines two main components: `QueryResultProvider` and `useQueryResult`.

Reviewer: ```suggestion
To help show how to avoid props-drilling, we'll first configure two components to use React Context: `QueryResultProvider` and `useQueryResult`.
```

interface QueryResult<TData> {
  data?: TData;
  error?: any;
  loading: boolean;
}

interface QueryContextValue<TData> {
  queryData: QueryResult<TData>;
  refetch: () => Promise<ApolloQueryResult<TData>>;
}

const QueryResultContext = createContext<QueryContextValue<any>>({
  queryData: { loading: true },
  refetch: () => Promise.resolve({} as ApolloQueryResult<any>),
});

interface QueryResultProviderProps<TVariables extends OperationVariables | undefined = {}> {
  query: any;
  variables?: TVariables;
  children: React.ReactNode;
}

export const QueryResultProvider = ({ query, variables, children }: QueryResultProviderProps) => {
  const { data, error, loading, refetch } = useQuery(query, { variables });
  const value = { queryData: { data, error, loading }, refetch };
  return (
    <QueryContext.Provider value={value}>{children}</QueryContext.Provider>
  );
}

export const useQueryResult = <TData = any>(): QueryResult<TData> => {
  const context = useContext(QueryResultContext);
  if (!context) {
    throw new Error(
      'useQueryResult must be used within a QueryResultProvider'
    );
  }
  return context.queryResult;
}
```

> In this example, the `QueryResultProvider` component is being used to wrap the `Users` component, providing it with the `GET_USERS` query. The `Users` component then uses the `useQueryResult` hook to access the data, loading state, and error from the query context.

Reviewer: ```suggestion
In this example, we use the `QueryResultProvider` component to wrap the `Users` component, providing it with the `GET_USERS` query. The `Users` component then uses the `useQueryResult` hook to access the data, loading state, and error from the query context.
```
